test/src/unit-cppapi-bitsort-filter.cc

```diff
@@ -0,0 +1,560 @@
+ * for a 3D array.
+ *
+ * @tparam DimType The type of the dimension.
+ * @param x_dims The first buffer that the function sets.
+ * @param y_dims The second buffer that the function sets.
+ * @param z_dims The third buffer that the function sets.
+ */
+template <typename DimType>
+void check_3d_dim_buffers(
+    const std::vector<DimType>& x_dims,
+    const std::vector<DimType>& y_dims,
+    const std::vector<DimType>& z_dims) {
+  size_t idx = 0;
+  for (int tile_idx_x = BITSORT_DIM_LO; tile_idx_x <= BITSORT_DIM_HI; tile_idx_x += TILE_EXTENT) {
+    for (int tile_idx_y = BITSORT_DIM_LO; tile_idx_y <= BITSORT_DIM_HI; tile_idx_y += TILE_EXTENT) {
+      for (int tile_idx_z = BITSORT_DIM_LO; tile_idx_z <= BITSORT_DIM_HI; tile_idx_z += TILE_EXTENT) {
+        for (int x = tile_idx_x; x < tile_idx_x + TILE_EXTENT; ++x) {
+          if (x > BITSORT_DIM_HI) {
+            break;
+          }
+          for (int y = tile_idx_y; y < tile_idx_y + TILE_EXTENT; ++y) {
+            if (y > BITSORT_DIM_HI) {
+              break;
+            }
+            for (int z = tile_idx_z; z < tile_idx_z + TILE_EXTENT; ++z) {
+              if (z > BITSORT_DIM_HI) {
+                break;
+              }
+
+              CHECK(x_dims[idx] == static_cast<DimType>(x));
+              CHECK(y_dims[idx] == static_cast<DimType>(y));
+              CHECK(z_dims[idx] == static_cast<DimType>(z));
+
+              idx += 1;
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+/**
+ * @brief Checks that the read query with row major layout returns the correct
+ * results for a 2D array.
+ *
+ * @tparam AttrType The type of the attribute of the array.
+ * @param global_a The attribute data in global order.
+ * @param a_data_read The attribute data in row-major order.
+ */
+template <typename AttrType>
+void check_2d_row_major(std::vector<AttrType>& global_a, std::vector<AttrType>& a_data_read) {
+  REQUIRE(global_a.size() == a_data_read.size());
+}
+
+/**
+ * @brief Checks that the read query with row major layout returns the correct
+ * results for a 3D array.
+ *
+ * @tparam AttrType The type of the attribute of the array.
+ * @param global_a The attribute data in global order.
+ * @param a_data_read The attribute data in row-major order.
+ */
+template <typename AttrType>
+void check_3d_row_major(std::vector<AttrType>& global_a, std::vector<AttrType>& a_data_read) {
+  REQUIRE(global_a.size() == a_data_read.size());
+}
+
+/**
+ * @brief Sets the subarrays to the entire array for a read query.
+ * Including this so we can test different code paths.
+ * 
+ * @tparam DimType The type of the dimension of the array.
+ * @param read_query The read query.
+ * @param num_dims The number of dimensions.
+ */
+template<typename DimType>
+void read_query_set_subarray(Query &read_query, int num_dims) {
+  DimType dim_lo = static_cast<DimType>(BITSORT_DIM_LO);
+  DimType dim_hi = static_cast<DimType>(BITSORT_DIM_HI);
+  read_query.add_range("x", dim_lo, dim_hi);
```

Comment:
I did not and this is why I wrote this code.

Reply:
My comment didn't show properly because github formatted it... I meant:

`read_query.add_range<DimType>("x", BITSORT_DIM_LO, BITSORT_DIM_HI);`

This builds fine on my machine.

Reply:
this worked for me too, thanks!